weld/codegen/llvm2/cmp.rs

```diff
@@ -0,0 +1,351 @@
+
+                // Finish block - set result and compute number of consumed bits.
+                LLVMPositionBuilderAtEnd(builder, done_block);
+                let result = LLVMBuildPhi(builder, self.i32_type(), c_str!(""));
+                let mut blocks = [entry_block, loop_block];
+                let mut values = [self.i32(1), cmp_result];
+                LLVMAddIncoming(result, values.as_mut_ptr(), blocks.as_mut_ptr(), values.len() as u32);
+
+                LLVMBuildRet(builder, result);
+            }
+            Function(_,_) | Unknown => {
+                unreachable!()
+            }
+        };
+
+        LLVMDisposeBuilder(builder);
+
+        self.cmp_fns.insert(ty.clone(), function);
+        Ok(function)
+    }
+
+    /// Generates an opaque comparison function for a type.
+    unsafe fn gen_opaque_cmp_fn(&mut self, ty: &Type) -> WeldResult<LLVMValueRef> {
+        let result = self.opaque_cmp_fns.get(ty).cloned();
+        if let Some(result) = result {
+            return Ok(result)
+        }
+
+        let llvm_ty = self.llvm_type(ty)?;
+        let mut arg_tys = [self.void_pointer_type(), self.void_pointer_type()];
+        let ret_ty = self.i32_type();
+
+        let c_prefix = LLVMPrintTypeToString(llvm_ty);
+        let prefix = CStr::from_ptr(c_prefix);
+        let prefix = prefix.to_str().unwrap();
+        let name = format!("{}.opaque_cmp", prefix);
+
+        // Free the allocated string.
+        LLVMDisposeMessage(c_prefix);
+
+        let (function, builder, _) = self.define_function_with_visibility(ret_ty,
+                                                                          &mut arg_tys,
+                                                                          LLVMLinkage::LLVMExternalLinkage,
+                                                                          name);
+
+        LLVMExtAddAttrsOnFunction(self.context, function, &[InlineHint]);
+        LLVMExtAddAttrsOnParameter(self.context, function, &[ReadOnly, NoAlias, NonNull, NoCapture], 0);
+        LLVMExtAddAttrsOnParameter(self.context, function, &[ReadOnly, NoAlias, NonNull, NoCapture], 1);
+
+        let left = LLVMGetParam(function, 0);
+        let right = LLVMGetParam(function, 1);
+        let left = LLVMBuildBitCast(builder, left, LLVMPointerType(llvm_ty, 0), c_str!(""));
+        let right = LLVMBuildBitCast(builder, right, LLVMPointerType(llvm_ty, 0), c_str!(""));
+
+        let func = self.gen_cmp_fn(ty)?;
+        let mut args = [left, right];
+        let result = LLVMBuildCall(builder, func, args.as_mut_ptr(), args.len() as u32, c_str!(""));
+        let result = LLVMBuildZExt(builder, result, self.i32_type(), c_str!(""));
+        LLVMBuildRet(builder, result);
+        LLVMDisposeBuilder(builder);
+
+        self.opaque_cmp_fns.insert(ty.clone(), function);
+        Ok(function)
+    }
+
+    unsafe fn gen_custom_cmp(&mut self,
+                             elem_ty: &LLVMTypeRef,
```

Comment:
nit: Since all the LLVM `Ref` types implement `Copy`, don't pass them around as references